zcash_client_backend/src/data_api/mem_wallet.rs

```diff
@@ -0,0 +1,459 @@
+#![allow(unused)]
+use incrementalmerkletree::Address;
+use secrecy::{ExposeSecret, SecretVec};
+use shardtree::{error::ShardTreeError, store::memory::MemoryShardStore, ShardTree};
+use std::{
+    cmp::Ordering,
+    collections::{BTreeMap, HashMap, HashSet},
+    convert::Infallible,
+    num::NonZeroU32,
+};
+use zcash_keys::keys::{AddressGenerationError, DerivationError};
+use zip32::{fingerprint::SeedFingerprint, DiversifierIndex};
+
+use zcash_primitives::{
+    block::BlockHash,
+    consensus::{BlockHeight, Network},
+    transaction::{Transaction, TxId},
+    zip32::AccountId,
+};
+use zcash_protocol::{
+    memo::{self, Memo, MemoBytes},
+    value::Zatoshis,
+};
+
+use crate::{
+    address::UnifiedAddress,
+    keys::{UnifiedAddressRequest, UnifiedFullViewingKey, UnifiedSpendingKey},
+    wallet::{NoteId, WalletTransparentOutput, WalletTx},
+};
+
+use super::{
+    chain::CommitmentTreeRoot, scanning::ScanRange, AccountBirthday, BlockMetadata,
+    DecryptedTransaction, NullifierQuery, ScannedBlock, SentTransaction, WalletCommitmentTrees,
+    WalletRead, WalletSummary, WalletWrite, SAPLING_SHARD_HEIGHT,
+};
+
+#[cfg(feature = "transparent-inputs")]
+use {crate::wallet::TransparentAddressMetadata, zcash_primitives::legacy::TransparentAddress};
+
+#[cfg(feature = "orchard")]
+use super::ORCHARD_SHARD_HEIGHT;
+
+struct MemoryWalletBlock {
+    height: BlockHeight,
+    hash: BlockHash,
+    block_time: u32,
+    // Just the transactions that involve an account in this wallet
+    transactions: HashMap<TxId, WalletTx<u32>>,
+    memos: HashMap<NoteId, MemoBytes>,
+}
+
+impl PartialEq for MemoryWalletBlock {
+    fn eq(&self, other: &Self) -> bool {
+        (self.height, self.block_time) == (other.height, other.block_time)
+    }
+}
+
+impl Eq for MemoryWalletBlock {}
+
+impl PartialOrd for MemoryWalletBlock {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some((self.height, self.block_time).cmp(&(other.height, other.block_time)))
+    }
+}
+
+impl Ord for MemoryWalletBlock {
+    fn cmp(&self, other: &Self) -> Ordering {
+        (self.height, self.block_time).cmp(&(other.height, other.block_time))
+    }
+}
+
+pub struct MemoryWalletAccount {
+    seed_fingerprint: SeedFingerprint,
+    account_id: AccountId,
+    ufvk: UnifiedFullViewingKey,
+    birthday: AccountBirthday,
+    addresses: BTreeMap<DiversifierIndex, UnifiedAddressRequest>,
+    notes: HashSet<NoteId>,
+}
+
+pub struct MemoryWalletDb {
+    network: Network,
+    accounts: BTreeMap<u32, MemoryWalletAccount>,
+    blocks: BTreeMap<BlockHeight, MemoryWalletBlock>,
+    tx_idx: HashMap<TxId, BlockHeight>,
+    sapling_spends: HashMap<sapling::Nullifier, (TxId, bool)>,
```

Comment:
How do you populate a field like this? I made some experiments and i always end up with an error:

```
the trait bound `sapling_crypto::Nullifier: Hash` is not satisfied
the trait `Hash` is not implemented for `sapling_crypto::Nullifier`
```

Reply:
Ah... apparently we need to add some impl's to the `sapling-crypto` crate. :sweat_smile:

Reply:
This is definitely a sketch at present.

Reply:
Ah, ok ,it will be ok to just derive it? https://github.com/zcash/sapling-crypto/pull/133

Reply:
Since `orchard` only derives `PartialOrd` and `Ord`, let's use BTreeMap instead (and add those to `sapling-crypto`).